onnx/helper.py

```diff
@@ -277,6 +277,16 @@ def split_complex_to_pairs(ca: Sequence[np.complex64]) -> Sequence[int]:
             for i in range(len(ca) * 2)]
 
 
+# convert a f32 (as uint32) to bf16 (as uint32)
+# TODO: big-endian
+# TODO: check for NaN before rounding to 'preserve' NaN payloads
+# algorithm:
+#  truncate bottom 16-bits, round remaining bits using round-to-nearest-even
```

Comment:
why we want this special handle for trancate? from bfloat16 definition in wiki, it seems just truncate the 16 bits from float32.
https://en.wikipedia.org/wiki/Bfloat16_floating-point_format

Reply:
When converting from FLOAT32 to BF16 truncation can work, since the encodings line up save for the 16 lest significant bits, but that means rounding errors. I don't want to get into the exact math, instead let me use a metaphor. let's say i want to convert a 3-digit number to a 2-digit number. Further, let's say the 3-digit number has the value `1.49`. if i convert via truncation i end up with `1.4` if i convert via _rounding_ i end up with `1.5`. The `1.5` value is closer to my original value and is "more correct". The rounding scheme i chose to implement is round-to-nearest-even. Which is the most common rounding scheme. Values below 0.5 are rounded down, values above 0.5 are rounded up, and values exactly equal to 0.5 are rounded to the nearest even value.

The comment i added here around NaN relates to what happens when the value isn't a number. For infinity values the lower 16-bits are 0s, so things work fine. but for NaN the lower mantissa may contain a "NaN payload", extra information that that value is carrying that further describes it beyond just NaN. Its probably not correct to round this value (since its not a number) but it also probably isn't correct to just truncate it. If detecting NaN was easy i would have not rounded it, but because i converted to integer value first, NaN is harder to detect.

Here is the code in pytorch that implements this: https://github.com/pytorch/pytorch/blob/e81bfffbe1349c5ba568a36ec911404c6e494ea0/c10/util/BFloat16.h#L51

Notice that pytorch uses round-to-nearest-even for converting from fp32 to bf16 and also it doesn't do rounding of NaN payloads.